cpp/serve/engine_actions/batch_verify.cc

```diff
@@ -0,0 +1,403 @@
+      RECORD_EVENT(trace_recorder_, requests[i]->id, "verify finish embedding");
+    }
+
+    RECORD_EVENT(trace_recorder_, request_ids, "start verify");
+    NDArray logits =
+        models_[verify_model_id_]->BatchVerify(embeddings, request_internal_ids, verify_lengths);
+    RECORD_EVENT(trace_recorder_, request_ids, "finish verify");
+    ICHECK_EQ(logits->ndim, 3);
+    ICHECK_EQ(logits->shape[0], 1);
+    ICHECK_EQ(logits->shape[1], total_verify_length);
+
+    std::vector<int> cum_verify_lengths = {0};
+    for (int i = 0; i < num_requests; ++i) {
+      cum_verify_lengths.push_back(cum_verify_lengths.back() + verify_lengths[i]);
+    }
+
+    // - Verify tokens.
+    logits = logits.CreateView({total_verify_length, 1, logits->shape[2]}, logits->dtype);
+    std::vector<RequestModelState> mstates;
+    std::vector<GenerationConfig> generation_configs;
+    int cur_req_id = 0;
+    for (int i = 0; i < total_verify_length; ++i) {
+      if (i >= cum_verify_lengths[cur_req_id + 1]) {
+        cur_req_id += 1;
+      }
+      Request request = requests[cur_req_id];
+      mstates.push_back(estate->GetRequestState(request)->mstates[verify_model_id_]);
+      generation_configs.push_back(request->generation_cfg);
+    }
+
+    logits = sampler_->ComputeProb(logits, models_[verify_model_id_], mstates, generation_configs);
+
+    ICHECK(logits.IsContiguous());
+    ICHECK(logits.DataType() == DataType::Float(32));
+
+    if (logits->device.device_type != kDLCPU) {
+      logits = logits.CopyTo(DLDevice{kDLCPU, 0});
+    }
+
+    ICHECK(logits->device.device_type == kDLCPU);
+
+    int64_t ndata = logits->shape[logits->ndim - 1];
+    const float* __restrict p_probs =
+        static_cast<float*>(__builtin_assume_aligned(logits->data, 4));
+
+    // this part is parallelizable
+    // TODO: parallelize this part
+    for (int i = 1; i <= num_requests; ++i) {
+      RequestModelState draft_mstate = rstates[i - 1]->mstates[draft_model_id_];
+      int verify_start = cum_verify_lengths[i - 1];
+      int verify_end = cum_verify_lengths[i];
+      int accept_length = VerifyAndAcceptTokens(
+          estate, requests[i - 1], p_probs, verify_start, verify_end, ndata,
+          draft_mstate->draft_output_tokens, draft_mstate->draft_output_token_prob,
+          draft_mstate->draft_output_prob_dist);
+      estate->stats.total_accepted_length += accept_length;
+      // minus one because the last draft token has no kv cache entry
+      int rollback_length = verify_end - verify_start - accept_length - 1;
+      // in case of all accepted
+      rollback_length = std::max(0, rollback_length);
+      // rollback kv cache
+      if (rollback_length > 0) {
+        models_[verify_model_id_]->PopNFromKVCache(
+            rstates[i - 1]->mstates[verify_model_id_]->internal_id, rollback_length);
+        models_[draft_model_id_]->PopNFromKVCache(
+            rstates[i - 1]->mstates[draft_model_id_]->internal_id, rollback_length);
+      }
```

Comment:
In case rollback happens (a.k.a. mismatch detected), shall we use `continue` to move on to the next request?

Reply:
The current logic is the same as using `continue`.